A model is required for all use cases when using agentic module

I want to use quarkus agent stuff, but only non-LLM agents and A2A. So I added the quarkus agentic dependency and the a2a langchain4j dependency, but when I start the app, I get

```
Caused by: io.quarkus.runtime.configuration.ConfigurationException: A ChatModel or StreamingChatModel bean was requested, but no langchain4j providers were configured. Consider adding an extension like 'quarkus-langchain4j-openai'
```

cc @mariofusco

--- a/agentic/deployment/src/main/java/io/quarkiverse/langchain4j/agentic/deployment/AgenticProcessor.java
+++ b/agentic/deployment/src/main/java/io/quarkiverse/langchain4j/agentic/deployment/AgenticProcessor.java
@@ -4,6 +4,7 @@ import static io.quarkiverse.langchain4j.agentic.deployment.ValidationUtil.valid
 import static io.quarkiverse.langchain4j.agentic.deployment.ValidationUtil.validateNoMethodParameters;
 import static io.quarkiverse.langchain4j.agentic.deployment.ValidationUtil.validateRequiredParameterTypes;
 import static io.quarkiverse.langchain4j.agentic.deployment.ValidationUtil.validateStaticMethod;
+import static io.quarkiverse.langchain4j.deployment.AiServicesProcessor.AGENTIC_PACKAGE_PREFIX;
 import static io.quarkiverse.langchain4j.deployment.AiServicesProcessor.TOOLBOX;
 
 import java.lang.reflect.InvocationHandler;
@@ -91,6 +92,14 @@ public class AgenticProcessor {
     private static final MethodDescriptor HANDLER_INVOKE = MethodDescriptor.ofMethod(
             InvocationHandler.class, "invoke", Object.class, Object.class, Method.class, Object[].class);
 
+    private static final Set<DotName> CHAT_MODEL_NOT_REQUIRED_ANNOTATIONS = Set.of(
+            AgenticLangChain4jDotNames.A2A_AGENT,
+            AgenticLangChain4jDotNames.SEQUENCE_AGENT,
+            AgenticLangChain4jDotNames.PARALLEL_AGENT,
+            AgenticLangChain4jDotNames.PARALLEL_MAPPER_AGENT,
+            AgenticLangChain4jDotNames.LOOP_AGENT,
+            AgenticLangChain4jDotNames.CONDITIONAL_AGENT);
+
     private static final List<DotName> ALL_CDI_CAPABLE_SUPPLIER_ANNOTATIONS = List.of(
             AgenticLangChain4jDotNames.CHAT_MODEL_SUPPLIER,
             AgenticLangChain4jDotNames.CHAT_MEMORY_SUPPLIER,
@@ -126,6 +135,9 @@ public class AgenticProcessor {
         nonAiMapProducer.produce(DetectedNonAiAgentAsMapBuildItem.from(classToNonAiAgentMethodsMap));
 
         ifaceToAgentMethodsMap.forEach((classInfo, methods) -> {
+            if (classInfo.name().toString().startsWith(AGENTIC_PACKAGE_PREFIX)) {
+                return;
+            }
             Optional<MethodInfo> chatModelSupplier = classInfo.methods().stream()
                     .filter(m -> Modifier.isStatic(m.flags()) && m.hasAnnotation(
                             AgenticLangChain4jDotNames.CHAT_MODEL_SUPPLIER))
@@ -418,7 +430,10 @@ public class AgenticProcessor {
 
     @BuildStep
     AnnotationsImpliesAiServiceBuildItem implyAiService() {
-        return new AnnotationsImpliesAiServiceBuildItem(AgenticLangChain4jDotNames.ALL_AGENT_ANNOTATIONS);
+        List<DotName> annotations = AgenticLangChain4jDotNames.ALL_AGENT_ANNOTATIONS.stream()
+                .filter(a -> !CHAT_MODEL_NOT_REQUIRED_ANNOTATIONS.contains(a))
+                .toList();
+        return new AnnotationsImpliesAiServiceBuildItem(annotations);
     }
 
     @BuildStep
@@ -608,14 +623,18 @@ public class AgenticProcessor {
         IndexView index = indexBuildItem.getIndex();
         Set<String> requestedChatModelNames = new HashSet<>();
         for (DetectedAiAgentBuildItem detectedAiAgentBuildItem : detectedAiAgentBuildItems) {
-            String chatModelName = detectedAiAgentBuildItem.getModelName() != null
-                    ? detectedAiAgentBuildItem.getModelName()
-                    : NamedConfigUtil.DEFAULT_NAME;
-            requestedChatModelNames.add(chatModelName);
-
-            AiAgentCreateInfo.ChatModelInfo chatModelInfo = detectedAiAgentBuildItem.getChatModelSupplier() != null
-                    ? new AiAgentCreateInfo.ChatModelInfo.FromAnnotation()
-                    : new AiAgentCreateInfo.ChatModelInfo.FromBeanWithName(chatModelName);
+            AiAgentCreateInfo.ChatModelInfo chatModelInfo;
+            if (detectedAiAgentBuildItem.getChatModelSupplier() != null) {
+                chatModelInfo = new AiAgentCreateInfo.ChatModelInfo.FromAnnotation();
+            } else if (requiresChatModel(detectedAiAgentBuildItem)) {
+                String chatModelName = detectedAiAgentBuildItem.getModelName() != null
+                        ? detectedAiAgentBuildItem.getModelName()
+                        : NamedConfigUtil.DEFAULT_NAME;
+                requestedChatModelNames.add(chatModelName);
+                chatModelInfo = new AiAgentCreateInfo.ChatModelInfo.FromBeanWithName(chatModelName);
+            } else {
+                chatModelInfo = new AiAgentCreateInfo.ChatModelInfo.NotNeeded();
+            }
 
             boolean hasInterceptorBindings = hasAnyInterceptorBindings(detectedAiAgentBuildItem, interceptorBindings, index);
             String ifaceName = detectedAiAgentBuildItem.getIface().name().toString();
@@ -736,6 +755,14 @@ public class AgenticProcessor {
             }
         }
         return false;
+    }
+
+    private static boolean requiresChatModel(DetectedAiAgentBuildItem agent) {
+        return agent.getAgenticMethods().stream()
+                .flatMap(m -> m.annotations().stream())
+                .map(AnnotationInstance::name)
+                .anyMatch(name -> AgenticLangChain4jDotNames.ALL_AGENT_ANNOTATIONS.contains(name)
+                        && !CHAT_MODEL_NOT_REQUIRED_ANNOTATIONS.contains(name));
     }
 
     @BuildStep
@@ -1218,7 +1245,7 @@ public class AgenticProcessor {
                 }
                 MethodInfo method = instance.target().asMethod();
                 // don't validate any of the upstream agents
-                if (method.declaringClass().name().toString().startsWith("dev.langchain4j.agentic")) {
+                if (method.declaringClass().name().toString().startsWith(AGENTIC_PACKAGE_PREFIX)) {
                     return;
                 }
                 List<MethodParameterInfo> parameters = method.parameters();

--- a/agentic/deployment/src/test/java/io/quarkiverse/langchain4j/agentic/deployment/validation/A2AOnlySequenceWithNoModelTest.java
+++ b/agentic/deployment/src/test/java/io/quarkiverse/langchain4j/agentic/deployment/validation/A2AOnlySequenceWithNoModelTest.java
@@ -1,0 +1,58 @@
+package io.quarkiverse.langchain4j.agentic.deployment.validation;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+import jakarta.enterprise.inject.CreationException;
+import jakarta.inject.Inject;
+
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import dev.langchain4j.agentic.declarative.A2AClientAgent;
+import dev.langchain4j.agentic.declarative.SequenceAgent;
+import dev.langchain4j.service.V;
+import io.quarkus.test.QuarkusUnitTest;
+
+/**
+ * Verifies that https://github.com/quarkiverse/quarkus-langchain4j/issues/2630 is fixed:
+ * a sequence of two A2A-only agents should not require a ChatModel bean to be configured.
+ */
+public class A2AOnlySequenceWithNoModelTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest unitTest = new QuarkusUnitTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
+                    .addClasses(StoryGenerator.class, StoryEditor.class, StoryPipeline.class));
+
+    @Test
+    void a2aOnlySequenceShouldNotRequireChatModel() {
+        // the exception is expected as there isn't any A2A server, but if we get there
+        // it means that the agents have been successfully configured, including the A2A one
+        assertThat(assertThrows(CreationException.class, () -> pipeline.run("dragons and wizards")))
+                .hasMessageContaining("Failed to obtain agent card");
+    }
+
+    @Inject
+    StoryPipeline pipeline;
+
+    public interface StoryGenerator {
+
+        @A2AClientAgent(a2aServerUrl = "http://localhost:9090", description = "Generate a story", outputKey = "story")
+        String generate(@V("topic") String topic);
+    }
+
+    public interface StoryEditor {
+
+        @A2AClientAgent(a2aServerUrl = "http://localhost:9091", description = "Edit a story for clarity", outputKey = "story")
+        String edit(@V("story") String story);
+    }
+
+    public interface StoryPipeline {
+
+        @SequenceAgent(outputKey = "story", subAgents = { StoryGenerator.class, StoryEditor.class })
+        String run(@V("topic") String topic);
+    }
+}

--- a/agentic/deployment/src/test/java/io/quarkiverse/langchain4j/agentic/deployment/validation/MissingChatModelWhenRequiredShouldFailTest.java
+++ b/agentic/deployment/src/test/java/io/quarkiverse/langchain4j/agentic/deployment/validation/MissingChatModelWhenRequiredShouldFailTest.java
@@ -1,0 +1,86 @@
+package io.quarkiverse.langchain4j.agentic.deployment.validation;
+
+import jakarta.inject.Inject;
+import jakarta.inject.Singleton;
+
+import org.assertj.core.api.Assertions;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import dev.langchain4j.agentic.Agent;
+import dev.langchain4j.agentic.declarative.A2AClientAgent;
+import dev.langchain4j.agentic.declarative.ChatModelSupplier;
+import dev.langchain4j.agentic.declarative.SequenceAgent;
+import dev.langchain4j.data.message.AiMessage;
+import dev.langchain4j.model.chat.ChatModel;
+import dev.langchain4j.model.chat.request.ChatRequest;
+import dev.langchain4j.model.chat.response.ChatResponse;
+import dev.langchain4j.service.UserMessage;
+import dev.langchain4j.service.V;
+import io.quarkus.test.QuarkusUnitTest;
+
+public class MissingChatModelWhenRequiredShouldFailTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest unitTest = new QuarkusUnitTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
+                    .addClasses(DummyChatModel.class, StoryCreator.class, CreativeWriter.class, AudienceEditor.class))
+            .assertException(throwable -> Assertions.assertThat(throwable)
+                    .hasMessageContaining("quarkus.langchain4j.openai.api-key"));
+
+    @Test
+    void test() {
+    }
+
+    @Singleton
+    public static class StoryCreatorConsumer {
+
+        @Inject
+        StoryCreator storyCreator;
+    }
+
+    public static class DummyChatModel implements ChatModel {
+
+        private final String response;
+
+        public DummyChatModel(String response) {
+            this.response = response;
+        }
+
+        @Override
+        public ChatResponse doChat(ChatRequest request) {
+            return ChatResponse.builder().aiMessage(new AiMessage(response)).build();
+        }
+    }
+
+    public interface CreativeWriter {
+
+        @A2AClientAgent(a2aServerUrl = "http://localhost:8080", description = "Generate a story based on the given topic", outputKey = "story")
+        String generateStory(@V("topic") String topic);
+
+        @ChatModelSupplier
+        static ChatModel chatModel() {
+            return new DummyChatModel("story");
+        }
+    }
+
+    public interface AudienceEditor {
+
+        @UserMessage("""
+                You are a professional editor.
+                Analyze and rewrite the following story to better align with the target audience of {{audience}}.
+                Return only the story and nothing else.
+                The story is "{{story}}".
+                """)
+        @Agent(description = "Edit a story to better fit a given audience", outputKey = "story")
+        String editStory(@V("story") String story, @V("audience") String audience);
+    }
+
+    public interface StoryCreator {
+
+        @SequenceAgent(outputKey = "story", subAgents = { CreativeWriter.class, AudienceEditor.class })
+        String write(@V("topic") String topic, @V("audience") String audience);
+    }
+}

--- a/agentic/runtime/src/main/java/io/quarkiverse/langchain4j/agentic/runtime/AgenticRecorder.java
+++ b/agentic/runtime/src/main/java/io/quarkiverse/langchain4j/agentic/runtime/AgenticRecorder.java
@@ -108,7 +108,8 @@ public class AgenticRecorder {
             @Override
             public Object apply(SyntheticCreationalContext<Object> cdiContext) {
                 ChatModel chatModel;
-                if (info.chatModelInfo() instanceof AiAgentCreateInfo.ChatModelInfo.FromAnnotation) {
+                if (info.chatModelInfo() instanceof AiAgentCreateInfo.ChatModelInfo.FromAnnotation
+                        || info.chatModelInfo() instanceof AiAgentCreateInfo.ChatModelInfo.NotNeeded) {
                     chatModel = null;
                 } else if (info.chatModelInfo() instanceof AiAgentCreateInfo.ChatModelInfo.FromBeanWithName b) {
                     if (NamedConfigUtil.isDefault(b.name())) {

--- a/agentic/runtime/src/main/java/io/quarkiverse/langchain4j/agentic/runtime/AiAgentCreateInfo.java
+++ b/agentic/runtime/src/main/java/io/quarkiverse/langchain4j/agentic/runtime/AiAgentCreateInfo.java
@@ -3,12 +3,16 @@ package io.quarkiverse.langchain4j.agentic.runtime;
 public record AiAgentCreateInfo(String agentClassName, ChatModelInfo chatModelInfo, boolean hasInterceptorBindings,
         boolean hasMcpToolBox) {
 
-    public sealed interface ChatModelInfo permits ChatModelInfo.FromAnnotation, ChatModelInfo.FromBeanWithName {
+    public sealed interface ChatModelInfo
+            permits ChatModelInfo.FromAnnotation, ChatModelInfo.FromBeanWithName, ChatModelInfo.NotNeeded {
 
         final class FromAnnotation implements ChatModelInfo {
         }
 
         record FromBeanWithName(String name) implements ChatModelInfo {
+        }
+
+        final class NotNeeded implements ChatModelInfo {
         }
     }
 }

--- a/core/deployment/src/main/java/io/quarkiverse/langchain4j/deployment/AiServicesProcessor.java
+++ b/core/deployment/src/main/java/io/quarkiverse/langchain4j/deployment/AiServicesProcessor.java
@@ -221,6 +221,8 @@ public class AiServicesProcessor {
 
     private static final ServiceOutputParser SERVICE_OUTPUT_PARSER = new QuarkusServiceOutputParser(); // TODO: this might need to be improved
 
+    public static final String AGENTIC_PACKAGE_PREFIX = "dev.langchain4j.agentic.";
+
     private static final Set<DotName> GUARDRAIL_ANNOTATIONS = Set.of(
             TOOL_INPUT_GUARDRAIL, TOOL_INPUT_GUARDRAILS, TOOL_OUTPUT_GUARDRAIL, TOOL_OUTPUT_GUARDRAILS);
 
@@ -673,6 +675,10 @@ public class AiServicesProcessor {
                     return;
                 }
                 if (!ci.isInterface()) {
+                    return;
+                }
+                if (ci.name().toString().startsWith(AGENTIC_PACKAGE_PREFIX)) {
+                    // Agentic interfaces and annotations are checked in the agentic module
                     return;
                 }
                 impliedDefaultRegisterAiService.add(AnnotationInstance.builder(REGISTER_AI_SERVICES).buildWithTarget(ci));

